File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
       if(head == null || head.next == null || k ==0 )return head;
       ListNode tail = head;int len = 1;
       while(tail.next != null){
        tail = tail.next;
        len++;
       } 
       tail.next = head;
       int st = len - k%len;
       ListNode Newtail = head;

       for(int i=1; i<st; i++) Newtail = Newtail.next;   
       ListNode newhead = Newtail.next;
       Newtail.next = null;

       return newhead;
    }
}

File: 61-rotate-list/rotate-list.py
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        length, tail = 1, head
        while tail.next:
            length += 1
            tail = tail.next
        k = k % length
        if k == 0:
            return head
        curr = head
        for _ in range(length - k - 1):
            curr = curr.next
        new_head, new_tail = curr.next, curr
        tail.next, new_tail.next = head, None
        return new_head